catalog/core/catalog-core-backupplugin/src/main/java/ddf/catalog/backup/PeriodicBatchExecutor.java

```diff
@@ -0,0 +1,133 @@
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class PeriodicBatchExecutor<T> {
+
+    List<T> items;
+
+    private ExecutorService executor;
+
+    private ScheduledExecutorService scheduledExecutor;
+
+    private Consumer<List<T>> task;
+
+    private int chunkSize;
+
+    private long period;
+
+    private TimeUnit timeUnit;
+
+    private BlockingQueue<Future> futures;
+
+    public PeriodicBatchExecutor(int chunkSize, long period, TimeUnit timeUnit) {
+        if (chunkSize < 1) {
+            throw new IllegalArgumentException("Parameter chunkSize must be greater than zero");
+        }
+        if (period < 1) {
+            throw new IllegalArgumentException("Parameter period must be greater than zero ");
+        }
+        this.chunkSize = chunkSize;
+        this.period = period;
+        this.timeUnit = timeUnit;
+        this.executor = Executors.newSingleThreadExecutor();
+        this.scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.futures = new LinkedBlockingQueue<>();
+        this.items = new ArrayList<>();
+    }
+
+    private synchronized void drain() {
+        for (List<T> batch : getBatches()) {
+            boolean success = futures.offer(executor.submit(() -> getTask().accept(batch)));
+            if(!success) {
+                throw new RuntimeException("Future object could not be added to the queue");
+            }
+        }
+    }
+
+    public synchronized void addAll(List<T> newItems) {
+        items.addAll(newItems);
+        if (isTankFull()) {
+            drain();
+        }
+    }
+
+    private boolean isTankFull() {
+        return items.size() >= chunkSize;
+    }
+
+    private void startSchedule() {
+        scheduledExecutor.scheduleAtFixedRate(this::drain, period, period, timeUnit);
+    }
+
+    private void shutdownExecutor() {
+        if (executor != null) {
+            executor.shutdown();
+        }
+    }
+
+    private void shutdownScheduledExecutor() {
+        if (scheduledExecutor != null) {
+            scheduledExecutor.shutdown();
+        }
+    }
+
+    public void shutdown() {
+        drain();
+        shutdownExecutor();
+        shutdownScheduledExecutor();
+    }
+
+    private Consumer<List<T>> getTask() {
+
+        return Objects.requireNonNull(task);
+    }
+
+    public void execute(Consumer<List<T>> command) {
+
+        this.task = Objects.requireNonNull(command);
+        startSchedule();
+    }
+
+    private List<List<T>> getBatches() {
+        List<List<T>> partitions = new ArrayList<>();
+        List<T> itemsCopy = new ArrayList<>(items);
+
+        for (int index = 0; index < itemsCopy.size(); index += chunkSize) {
```

Comment:
Should be possible to simplify this quite a bit using http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Lists.html#partition.